DO NOT MERGE: system table rbr integration branch

jeffswenson opened the pull request.

jeffswenson committed:
sqlliveness: migrate to an rbr compatible index

- update the bootstrap system schema to use the new index
- add version guards for the mgiration
- add upgrade tasks to backfill the index and clean up the old index

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/97406)
<!-- Reviewable:end -->

JeffSwenson commented:
@ajwerner This is not quite ready. I still need to write a lot of tests. Can you take a look at the overall approach and let me know if you think it needs to change?

jeffswenson committed:
sqlinstance: migrate to rbr compatible table

<what was there before: Previously, ...>
<why it needed to change: This was inadequate because ...>
<what you did about it: To address this, this patch ...>

jeffswenson committed:
upgrade: add system rbr upgrade and versions

This commit contains the version and upgrade tasks needed to migrate the
system.sql_instances, system.lease, and system.sqlliveness tables to an
index format compatible with regional by row tables.

Follow on work is needed to add version checks to code interacting with
the tables. The migrations will be enabled for each table as the version
checks are implemented. The draft PR #97406 contains version checks for
each of the tables and was manually tested with this change to ensure
the upgrade process works as expected.

Part of #98111

jeffswenson committed:
sqlinstance: refactor sql instance reader

This change is intended to support the migration to a regional by row
compatible index structure. Encapsulating the feed into a sub object
makes it easy to swap the feed implementation when the index changes.

Release note: None

jeffswenson committed:
sqlinstance: more work on rbr refactor

ajwerner commented:
Sorry for the slow turnaround. The idea seems good to me.

jeffswenson committed:
lease: migrate lease to rbr compatible table

jeffswenson committed:
server: move settingswatcher initialization to the start

Previously, the settingswatcher was one of the last services initialized
during sql server start up. Now, it is one of the first services to
start up. The settings do not have a well defined value until after
settingwatcher initializes. Some settings, like cluster version, must
have valid values during initialization.

Part of #94843

jeffswenson committed:
upgrades: enable rbr upgrades

<what was there before: Previously, ...>
<why it needed to change: This was inadequate because ...>
<what you did about it: To address this, this patch ...>

JeffSwenson commented:
Closing this PR now that all of the changes are merged
